jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/WebSocketDemander.java

```diff
@@ -75,18 +83,54 @@ public DemandingFlusher(IncomingFrames emitFrame)
      */
     protected abstract boolean handle(Frame frame, Callback callback, boolean first);
 
-    @Override
-    public void demand()
+    /**
+     * <p>This is used within an implementation of {@link #handle(Frame, Callback, boolean)}
+     * to forward a frame onto the next layer of processing.</p>
+     *
+     * <p>This method should only be called ONCE within each invocation of {@link #handle(Frame, Callback, boolean)}
+     * otherwise</p>
+     * @param frame the WebSocket frame.
+     * @param callback to release frame payload.
+     */
+    public void emitFrame(Frame frame, Callback callback)
     {
-        if (!_demand.compareAndSet(false, true))
-            throw new ReadPendingException();
-        iterate();
+        Throwable failure = null;
+        try (AutoLock l = _lock.lock())
+        {
+            if (!_downstreamDemand)
+                failure = new IllegalStateException("Demand already fulfilled");
+            else
+                _downstreamDemand = false;
+        }
+
+        if (failure != null)
+        {
+            callback.failed(failure);
+            failFlusher(failure);
+        }
+        else
+        {
+            _emitFrame.onFrame(frame, callback);
+        }
     }
 
+    /**
+     * This is used by the downstream layer to demand a new frame to be processed in {@link #handle(Frame, Callback, boolean)}.
```

Comment:
Oh - I would have called the layer calling demand upstream, as it is the layer above.
Hmmmm maybe I suggested bad names for you???

Reply:
I actually think you meant upstream

Reply:
The fact that we need to have this discussion probably means the names aren't suitable and are likely to cause some confusion. 

I can try using producerDemand and consumerDemand names.

>  Also you have used "Next" in at least once circumstance

This is a pre-existing method on the `DemandChain` interface. But it would be a little confusing to rename this to `setProducerDemand()` when there is no concept of consumer demand on the interface.